### ISSUE TITLE
[FR] Passing any JavaScript object/class to AssemblyScript

### ISSUE THREAD

**Issue by aminya** [AssemblyScript/assemblyscript#1620]
It would be nice if AssemblyScript adds support for passing any JavaScript object/class to WASM using variants. This is similar to how node-addon-api support passing any JavaScript object as [Napi::Object](https://github.com/nodejs/node-addon-api/blob/master/doc/object.md#example). 

 The concentrate is to make it easier to pass **random** objects to AssemblyScript without the need for writing the types for it (the efficiency is not a priority).  

The proposed syntax
```ts
function asFunc(obj: JS.Object) {

	// getting values
	const x = obj.get(1) as string
	
	// setting values
	obj.set(2, 1)
}
```

As a bonus, optional typing might be provided for a subset of the object that is needed:
```ts
// a subset of the object which is typed
interface MyObject {
	desiredMethod(a: i32): i32
}

function asFunc(obj: JS.Object) {
	const x = obj.get(1) as MyObject
	x.desiredMethod()
}
```

This is different from the reference-type proposal, which requires strong typing of the passed object (more efficient).

The possible implementation will look like this:
https://github.com/nearprotocol/assemblyscript-json/blob/master/assembly/JSON.ts#L103
https://github.com/nearprotocol/assemblyscript-json/blob/master/assembly/JSON.ts#L202

P.S: I already discussed this with @MaxGraey in the discord chat.

**Comment by dcodeIO** [comment]
How would the object be referenced from within Wasm, if not with a reference type? And how would the compiler determine the intention of the programmer, to generate the necessary typed imports to interface with the object, if there are no types?

**Comment by aminya** [comment]
Emscripten already does it:
https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#using-val-to-transliterate-javascript-to-c

**Comment by MaxGraey** [comment]
I already suggested this in Discord and duplicating here. With reference types it will be much more powerful and simpler. Something like:
```ts
@external("webapi", "DOMNode")
interface DOMNodeExternal {
  replaceChild(oldNode: DOMNodeExternal, newNode: DOMNodeExternal): DOMNodeExternal;
  ...
}

type DOMNode = Externref<DOMNodeExternal>

export function replaceChild(oldChild: DOMNode, _newChild: DOMNode): DOMNode {
  return old.replaceChild(oldChild, _newChild);
}
```
And you don't need expensive Variant accessor. But reftypes (externref) still very experimental and required more adoption from compiler for that

**Comment by aminya** [comment]
> But reftypes (externref) still very experimental and required more adoption from compiler for that

Can we have something less efficient for now so we can start thinking about using WASM? Without this, it is impossible for me to port some of my apps to WASM.

If we choose a good syntax, the same thing can be used for reference types which makes the later transition to actual reference types as easy as changing a compiler flag.

**Comment by trusktr** [comment]
> Can we have something less efficient for now, so we can start thinking about using WASM?

Of course it's possible, but someone has to spend time on it. That's the main problem. But I agree, it would provide a way to start writing code now that would get a performance boost from reference types later, without existing user code changing much (if at all).

Can it be done as a library for AS (not in core)?

Whatever the solution will be, I hope we can keep it valid TypeScript syntax. In the previous example,

```js
@external("webapi", "DOMNode")
interface DOMNodeExternal {
  replaceChild(oldNode: DOMNodeExternal, newNode: DOMNodeExternal): DOMNodeExternal;
  ...
}
```

that is invalid TypeScript, because decorators are not allowed there.

**Comment by aminya** [comment]
Unstale

**Comment by spikewoodman** [comment]
I have a related requirement;
https://stackoverflow.com/questions/74554478/passing-javascript-references-to-assemblyscript

**Comment by snowyu** [comment]
FYI: https://github.com/web-devkits/Wasmnizer-ts can compile ts to wasm directly. support any type.

**Comment by JairusSW** [comment]
@snowyu, Wasmnizer relies on external imports, so it can only run in a few select environments

**Comment by snowyu** [comment]
Is this a very difficult feature?  4 years have passed since the issue was published, and it still hasn't been implemented. JS Object can only be passed through inefficient JSON string.

**Comment by JairusSW** [comment]
@snowyu, you can pass objects through linear memory

**Comment by snowyu** [comment]
> pass objects through linear memory
Yes, decode/encode jsobject is very hard?

Can find nothing about assemblyscript externref:

* rust: https://github.com/slowli/externref
* c: https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/doc/ref_types.md

**Comment by JairusSW** [comment]
@snowyu, can you hop over to the discord server? I can help you better there

**Comment by CountBleck** [comment]
The issue with `externref` is that they can't currently be non-static class members, because `externref`s can't be stored in linear memory. (Unless...you have a wrapper `class` that stores them to a dedicated table and that has a finalizer to remove the `externref` from the table...I could make a simple transform/library demo for that if needed?)

**Comment by snowyu** [comment]
Thanks all. I just wanna find a better lazy way to pass the js dynamic object into wasm and return the js object(dynamic):

```ts
async function exec(args: any): any
```

**Comment by JairusSW** [comment]
Do you really need to have it accept plain objects? What's your use case?

Like yeah, you can use JSON, but you can't even deserialize an arbitrary object so it doesn't even make sense

**Comment by JairusSW** [comment]
And yeah if you want to pass objects around id still use externref

**Comment by snowyu** [comment]
It's total dynamic. in fact the exec function just a router which will call other functions in wasm.

**Comment by JairusSW** [comment]
https://www.assemblyscript.org/compiler.html#host-bindings

**Comment by snowyu** [comment]
I do not know your meaning. Yes, the assemblyscript has PlainObject/ComplexObject, but they are also static.

The possible resolution is to implement the [variant](https://github.com/MaxGraey/as-variant), and connect to [as-bind](https://github.com/torch2424/as-bind/).

**Comment by JairusSW** [comment]
As-bind isn't used anymore. I'd use the generated bindings and variants

**Comment by snowyu** [comment]
> As-bind isn't used anymore. I'd use the generated bindings and variants

Can it be generated dynamically at runtime?

**Comment by JairusSW** [comment]
No, but with the --exposeRuntime flag you can pass high level data types manually
